### PR TITLE
[PoC] Launch test YAMLs directly

### PR DESCRIPTION
**Pull request by josvazg**
PoC  not to merge but to be used for API testing.

### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

### CODE REVIEW

File: test/helper/launcher/launcher.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	if !ok {
+		panic(fmt.Errorf("environment variable %s not set", envvar))
+	}
+	return value
+}
+
+// Launch will try to launch the operator and apply the given YAML for it to handle
+func (l *Launcher) Launch(yml string, waitCfg *WaitConfig) error {
+	if err := l.ensureK8sCluster(); err != nil {
+		return err
```
Comment (s-urbaniak) [inline test/helper/launcher/launcher.go:57]: wrap all errors below. In this case:
```suggestion
		return fmt.Errorf("error ensuring k8s cluster: %w", err)
```

**Comment by s-urbaniak** [comment]
See https://kuttl.dev/docs/kuttl-test-harness.html#writing-your-first-test and https://kuttl.dev/docs/api-integration.html#test-harness.

**Comment by josvazg** [comment]
> See https://kuttl.dev/docs/kuttl-test-harness.html#writing-your-first-test and https://kuttl.dev/docs/api-integration.html#test-harness.

This is not a testing framework, it is a simple testing helper. It could be used as a framework maybe, but it is NOT the intent here and for that I agree that KUTTL is a better option.

Think of this as the function I call in Go to get the Atlas infrastructure in the state I need to run a contract test, which is defined in Go code, not in YAML. A test that calls our translation layer that talks to the API using the same code that the reconcile logic uses. If KUTTL could be used in the same unobtrusive way, then maybe we could use it for all, I am not 100% convinced yet.

Sample custom test:

```go
  l := launcher.NewFromEnv(testVersion) // instantiate a launcher
  l.Launch(testYml, launcher.WaitReady("atlasprojects/my-project", 30*time.Second))) // launch the YAML and wait for it
  
  // run custom test code - might not be part of KUTTLE, like a custom go code driven contract test

   l.Cleanup()) // optionally cleanup the infrastructure required for the test (does KUTTL allow for optional cleanup?)
```

**Comment by s-urbaniak** [comment]
@josvazg I am leaving the final assertion to you 👍 I didn't have a deeper look at kuttl, just worried about NIH implementations. Again, final call is yours :-)

File: test/helper/launcher/wait.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	Timeout   time.Duration
+}
+
+func WaitReady(target string, timeout time.Duration) *WaitConfig {
+	return &WaitConfig{Condition: "condition=Ready", Target: target, Timeout: timeout}
+}
+
+func (cfg *WaitConfig) waitArgs() []string {
+	args := []string{"wait", fmt.Sprintf("--for=%s", cfg.Condition), cfg.Target}
+	if cfg.Timeout == 0 {
```
Comment (helderjs) [inline test/helper/launcher/wait.go:24]: It's a mistake or do you really want to set timeout only when it's zero?

**Comment by josvazg** [comment]
> @josvazg I am leaving the final assertion to you 👍 I didn't have a deeper look at kuttl, just worried about NIH implementations. Again, final call is yours :-)

I think the best way forward now is to actually force myself to use KUTTL for some real Contract tests, for Atlas Search. If I see no blockers, then we can probably use for everything and forget about this PoC for good.

**Review by helderjs — CHANGES_REQUESTED** [review]

**Comment by helderjs** [comment]
@josvazg This PoC links well to my comment on the CI Revamp document regarding GitHub Actions.

Why didn't you write a Github action workflow for it? I'd say we probably won't need any code for it.
I see your code doing what Github Action should be doing, and I believe that's why @s-urbaniak also saw it as a framework wrapping the tooling needed to achieve the goal. Code is a liability and I prefer to have no code when we already have a toolchain in place to do the job. In the end, you would still use a workflow to trigger this task.
Think that a CI tool like GH action offers us out-of-box multi-stage pipeline and jobs, some level of security, lifecycle control... keep in mind that once you have signaled you will drop this to a new PoC with KUTTL

**Comment by josvazg** [comment]
> @josvazg This PoC links well to my comment on the CI Revamp document regarding GitHub Actions.
> 
> Why didn't you write a Github action workflow for it? I'd say we probably won't need any code for it. I see your code doing what Github Action should be doing, and I believe that's why @s-urbaniak also saw it as a framework wrapping the tooling needed to achieve the goal. Code is a liability and I prefer to have no code when we already have a toolchain in place to do the job. In the end, you would still use a workflow to trigger this task. Think that a CI tool like GH action offers us out-of-box multi-stage pipeline and jobs, some level of security, lifecycle control... keep in mind that once you have signaled you will drop this to a new PoC with KUTTL

I want us to be able to do our development without any Github involvement, using the github CI actions only to automatically re-run tests at the proper time to ensure what we release is correct and did not include any regressions.

The idea is to be able to write contract tests in Go that look like this:

1. Get a particular set of Atlas resources in a particular state.
1. Run one or more contract tests against the Atlas API.
1. Tear down all resources required for the test.

During development, in your laptop and before any code reaches github, you iterate with steps 1-2 until you test all APIs you needed to verify. Then do step 3. It is only later on you send a PR and then the CI might also want to run those tests at some point.

This might be doable with KUTTL, because it seems to have a programmatic interface. I will need to test it.
But as a pure github action or workflow it would be completely useless for contract testing.

**Comment by josvazg** [comment]
Going to close this PoC now:
- It was never the intention to merge this at this point.
- It was needed for me as a way to run contract tests, but haven't gotten there yet.
- Seems KUTTL has a programmatic interface that might allow it to be used within tests as a helper, in which case this code will not be needed.
- Should KUTTL prove inadequate later, we can always dig up this Pr again.
- The main goal of the PoC is achieved, launching declarative YAML infrastructure for a test OR as a helper to setup before a test is not only doable but also relatively simple.

File: test/helper/launcher/wait.go (hunk 2 above)
Comment (josvazg) [reply test/helper/launcher/wait.go:24]: This was a mistake, the logic had to be reversed

**Pull request closed without merging**